pkg/agent/flowexporter/connections/conntrack_linux.go

```diff
@@ -17,61 +17,83 @@
 	"k8s.io/klog"
 
-	"github.com/vmware-tanzu/antrea/pkg/agent/config"
 	"github.com/vmware-tanzu/antrea/pkg/agent/flowexporter"
 	"github.com/vmware-tanzu/antrea/pkg/agent/openflow"
 	"github.com/vmware-tanzu/antrea/pkg/agent/util/sysctl"
+	"github.com/vmware-tanzu/antrea/pkg/ovs/ovsconfig"
+	"github.com/vmware-tanzu/antrea/pkg/ovs/ovsctl"
+	"github.com/vmware-tanzu/antrea/pkg/util/ip"
 )
 
-var _ ConnTrackDumper = new(connTrackDumper)
-
-type connTrackDumper struct {
-	nodeConfig  *config.NodeConfig
-	serviceCIDR *net.IPNet
-	connTrack   ConnTrackInterfacer
-}
-
-func NewConnTrackDumper(nodeConfig *config.NodeConfig, serviceCIDR *net.IPNet, conntrack ConnTrackInterfacer) *connTrackDumper {
-	return &connTrackDumper{
-		nodeConfig,
-		serviceCIDR,
-		conntrack,
-	}
-}
-
 // DumpFlows opens netlink connection and dumps all the flows in Antrea ZoneID
 // of conntrack table, i.e., corresponding to Antrea OVS bridge.
 func (ctdump *connTrackDumper) DumpFlows(zoneFilter uint16) ([]*flowexporter.Connection, error) {
-	// Get netlink Connection to netfilter
-	err := ctdump.connTrack.Dial()
-	if err != nil {
-		klog.Errorf("Error when getting netlink conn: %v", err)
-		return nil, err
+	if ctdump.datapathType == ovsconfig.OVSDatapathSystem {
+		// Get connection to netlink socket
+		err := ctdump.connTrack.GetConnTrack(nil)
+		if err != nil {
+			klog.Errorf("Error when getting netlink conn: %v", err)
+			return nil, err
+		}
+	} else if ctdump.datapathType == ovsconfig.OVSDatapathNetdev {
+		// Set ovsCtlClient to dump conntrack flows
+		err := ctdump.connTrack.GetConnTrack(ctdump.ovsctlClient)
+		if err != nil {
+			klog.Errorf("Error when getting ovsclient: %v", err)
+			return nil, err
+		}
 	}
 
 	// ZoneID filter is not supported currently in tl-mo/conntrack library.
 	// Link to issue: https://github.com/ti-mo/conntrack/issues/23
 	// Dump all flows in the conntrack table for now.
-	conns, err := ctdump.connTrack.DumpFilter(conntrack.Filter{})
-	if err != nil {
-		klog.Errorf("Error when dumping flows from conntrack: %v", err)
-		return nil, err
+	var conns []*flowexporter.Connection
+	var err error
+	if ctdump.datapathType == ovsconfig.OVSDatapathSystem {
+		conns, err = ctdump.connTrack.DumpFilter(conntrack.Filter{})
+		if err != nil {
+			klog.Errorf("Error when dumping flows from conntrack: %v", err)
+			return nil, err
+		}
+	} else if ctdump.datapathType == ovsconfig.OVSDatapathNetdev {
+		// This is supported for kind clusters. Ovs-appctl access in kind clusters is unstable currently.
+		// This will be used once the issue with Ovs-appctl is fixed on kind cluster nodes.
+		conns, err = ctdump.connTrack.DumpFilter(uint16(openflow.CtZone))
+		if err != nil {
+			klog.Errorf("Error when dumping flows from conntrack: %v", err)
+			return nil, err
+		}
 	}
 
-	filteredConns := make([]*flowexporter.Connection, 0, len(conns))
-	for _, conn := range conns {
-		if conn.Zone != openflow.CtZone {
+	for i := 0; i < len(conns); i++ {
```

Comment:
I'm not sure if there is a performance reason for this, but the code was much more readable when you were building a new slice IMO.